### ISSUE TITLE
MinResults and MaxResults in Search() function implemented improperly, Doesn't allow for getting every listing available.

### ISSUE THREAD

**Issue by DudeManGuy0** [mwpenny/kijiji-scraper#47]
Currently there is no real way to get all the available listings using MinResult and MaxResult. You'd need to set MinResult to something like 60 and if there are only 59 listings, it will error with this message:
`Error parsing Kijiji search results: Result ad has no URL. It is possible that Kijiji changed their markup. If you believe this to be the case, please open an issue at: https://github.com/mwpenny/kijiji-scraper/issues`

With the way MinResults works, you specify the minimum and you get no more. While the way it should work is say setting MinResults to 40 and MaxResults to 50 and getting 50. This allows for all the functionality of selecting how many listings you can get and also getting as many as there is.

To summarize, even if you set MaxResults to -1 indicating no limit, you will only receive how many MinResults is rounded every 20. Instead it should get as many listings as it can before reaching the MaxLimit. This way you can set MinResults to 20 and MaxResults to 60 and get 60 listings.

**Comment by mwpenny** [comment]
You're misunderstanding what `maxResults` is for. Per the readme, the purpose of `maxResults` is not for specifying how many results you will receive but rather for _limiting_ how many results you will receive. The reason it works this way is because search result pages have a minimum number of ads. So if you specify `minResults` as `1`, for example, you will usually get many more. The `maxResults` functionality is just a simple wrapper around `Array.slice()` to discard extra results you're not interested in so you don't have to do it yourself.

In short, when you search, you will get at least `minResults` and at most `maxResults` search results. I suppose I could implement `-1` for `minResults` in order to mean "no limit". 

As for the error you're talking about - that should never happen unless Kijiji changed something. Can you provide reproduction steps? That sounds like a real issue.

**Comment by DudeManGuy0** [comment]
You can have the functionality you describe as well as what I have. You can have maxResults to specify how many you need and it will trim off the excess to get how many you asked for. You don't need 2 variables for this and can make it work with just one while having the option to get all listings. Whether you want to implement this design is your choice, in my eyes it's better though.

To reproduce the issue, make a request with q, locationId,categoryId, and set minResults to something like 500 (or any high amount) and set maxResults to -1. It should bring the error I described. Here is the code:

```
const kijiji = require("kijiji-scraper");

const options = {
    minResults: 20000,
    maxResults: -1
};

const params = {
    q: "iphone",
    locationId: 1700199,  // Same as kijiji.locations.ONTARIO.OTTAWA_GATINEAU_AREA.OTTAWA
    categoryId: 0,  // Same as kijiji.categories.CARS_AND_VEHICLES
    sortByName: "priceAsc"  // Show the cheapest listings first
};

kijiji.search(params, options).then(ads => {
    for (let i = 0; i < ads.length; ++i) {
        console.log(ads[i].attributes.price)
        console.log(ads[i].title);
        console.log(ads[i].description)
        console.log("\n")
    }
}).catch(console.error);
```
However, set `minResults` to 20 and it will work perfect.

**Comment by mwpenny** [comment]
It's technically possible to make it work with just one, but the reason there are two separate variables is to support the use case where you don't care if you get more results than you asked for. Your proposed change to `maxResults` would change its meaning. In addition to specifying the maximum number of results, it would also be implicitly specifying the minimum as well.

I think there's a middle ground to be had here since I can see where you're coming from. I plan to make the following changes:
* A negative value for `minResults` will mean "no limit"
* `minResults` will take on the value of `maxResults` if:
    1. `minResults === undefined` and `maxResults > 0`, or
    2. `minResults` is negative
* If both `minResults` and `maxResults` are unspecified, the current behavior is unchanged (`minResults = 20`).

This keeps the current behavior by default but allows for optional use of the `maxResults` behavior that you originally described as well as "no limit" searching. How does that sound?

**Comment by mwpenny** [comment]
Also, I was able to reproduce your issue. It doesn't have anything to do with `minResults` or `maxResults`. It's because some ads returned by the API do not have an associated URL on the public-facing Kijiji website (some "top ads"). I've pushed a fix that just skips such ads (#48). Please try again with the latest version.

**Comment by DudeManGuy0** [comment]
I can see where you were coming from as well I guess. We both have different views for what minResults and maxResults should be and their use cases. I'd be fine with that solution as my only aim is to be able to get all listings available. I'll try your solution to the problem I posted earlier too. Another improvement you might want to implement is detection for ads from companies so you can easily sort between real listings and ads from companies which aren't desired. The reason is usually they are unrelated to what you are searching for. This one is your choice though as it doesn't really bother me and probably doesn't bother anyone else. Thank you for your active contribution to this repository.

**Comment by mwpenny** [comment]
Alright, sounds good. I'll merge the new behavior when I get some holiday downtime. I'd like to test it more first against the real servers.

Also, I'm all for filtering out third-party ads. Back when this module was only scraping the HTML of the public website, I outright excluded third-party ads for the reasons you're describing. They're typically pretty spammy and sometimes link to non-Kijiji websites (which is out of the scope of this project). When I switched to the mobile API by default, I forgot to filter those. I think that my fix to the "no URL" issue should help with them. Let me know if you're getting these types of ads in your results and I can look into a better way of dealing with them - it should be easier with the API scraper than with the HTML scraper.

**Comment by mwpenny** [comment]
Glad to see people are getting good use out of this!

**Comment by DudeManGuy0** [comment]
Did you implement setting MinResults to -1 and getting all listings available yet?

**Comment by mwpenny** [comment]
Hi there, I got some more time to look at this and have merged the changes discussed above. Version 6.1.0 (now on NPM) has these changes. Let me know if you run into problems.

All the best!
